### ISSUE TITLE
MA0062 - Non-flags enums should not be marked with FlagsAttribute is handled by CA2217

### ISSUE THREAD

**Issue by Meir017** [meziantou/Meziantou.Analyzer#850]
- **Rule**: [MA0062](https://github.com/meziantou/Meziantou.Analyzer/blob/main/docs/Rules/MA0062.md) - Non-flags enums should not be marked with "FlagsAttribute"
- **Built-in Equivalent**: [CA2217](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2217) - Do not mark enums with FlagsAttribute

## Analysis

**MA0062** and **CA2217** serve the same purpose: ensuring that the `FlagsAttribute` is only applied to enums that are designed for bitwise operations, preventing misuse of the attribute on regular enums.

### MA0062 Coverage
- Detects FlagsAttribute on non-flags enums
- Ensures proper enum design patterns
- Prevents incorrect bitwise operations

### CA2217 Coverage  
- Detects inappropriate use of FlagsAttribute on enums
- Official Microsoft rule for proper enum design
- Validates that FlagsAttribute is used correctly
- Part of the .NET design guidelines

## Code Examples

**Problematic code detected by both rules:**
```csharp
// Both MA0062 and CA2217 will flag this
[Flags] // Incorrect - not designed for bitwise operations
public enum Status
{
    Pending = 1,
    InProgress = 2,
    Completed = 3,    // Not a power of 2!
    Failed = 4
}

[Flags] // Incorrect - sequential values
public enum Priority
{
    Low = 1,
    Medium = 2,
    High = 3,         // Not a power of 2!
    Critical = 4
}
```

**Recommended fix:**
```csharp
// Option 1: Remove FlagsAttribute for simple enums
public enum Status
{
    Pending,
    InProgress,
    Completed,
    Failed
}

// Option 2: Design properly for flags if bitwise operations are needed
[Flags]
public enum FilePermissions
{
    None = 0,
    Read = 1,         // 2^0
    Write = 2,        // 2^1
    Execute = 4,      // 2^2
    ReadWrite = Read | Write,
    All = Read | Write | Execute
}
```

## Design Issues Prevented

Incorrect FlagsAttribute usage can cause:
- **Unexpected bitwise behavior** when combining enum values
- **Logical errors** in flag checking operations
- **ToString() formatting issues** - shows combined values incorrectly
- **API confusion** - misleads consumers about intended usage

## Recommendation

**CA2217** is the official Microsoft analyzer rule that provides **equivalent functionality** to MA0062. Since both rules target the same enum design issue and CA2217 is the standard rule for enum validation, **MA0062 should be deprecated** in favor of CA2217.

## Migration Path

1. Enable **CA2217** in your analyzer configuration:
   ```xml
   <WarningsAsErrors Include="CA2217" />
   ```
2. Disable **MA0062**:
   ```xml
   <WarningsNotAsErrors Include="MA0062" />
   ```
3. Update any rule suppressions from MA0062 to CA2217

## Related

- Parent issue: #835 - Consider deprecating rules that are now covered by built-in .NET analyzers

**Comment by Meir017** [comment]
## Analysis: MA0062 vs CA2217 - Equivalent Rules ✅

**MA0062** and **CA2217** serve the same purpose: preventing incorrect `FlagsAttribute` usage on non-flags enums.

### Recommendation: Deprecate MA0062 ✅

**CA2217** is the official Microsoft rule for enum design validation. **MA0062 should be deprecated** in favor of CA2217.

---
*Part of #835 deprecation initiative.*

**Comment by Meir017** [comment]
This issue should be closed - MA0056 equivalency is already documented in [docs/comparison-with-other-analyzers.md](https://github.com/meziantou/Meziantou.Analyzer/blob/main/docs/comparison-with-other-analyzers.md) - MA0056 is equivalent to CA2214.

**Comment by Meir017** [comment]
This mapping is already documented in the [comparison file](https://github.com/meziantou/Meziantou.Analyzer/blob/main/docs/comparison-with-other-analyzers.md#equivalent-rules) under "Equivalent rules":

| [CA2217](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2217?WT.mc_id=DT-MVP-5003978) | [MA0062](https://github.com/meziantou/Meziantou.Analyzer/blob/main/docs/Rules/MA0062.md) |

After analyzing both rules, they are indeed **equivalent**:

## Functional Equivalence
- Both detect `[Flags]` enums with non-power-of-2 values
- Both use nearly identical examples (Color enum with Orange = 3)
- Both aim to ensure proper enum design for bitwise operations
- Both detect the same violations and have equivalent functionality

## Minor Differences
- **MA0062** has a specific configuration option for `All = ~None` patterns
- **CA2217** has API surface configuration options
- **CA2217** is the official Microsoft rule

Since CA2217 provides equivalent coverage and is the official Microsoft analyzer rule, the mapping is correctly documented as "equivalent rules" rather than needing any additional coverage analysis.
